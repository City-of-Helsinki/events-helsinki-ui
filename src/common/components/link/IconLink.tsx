import classNames from "classnames";
import React from "react";
import { Link, LinkProps } from "react-router-dom";

import styles from "./iconLink.module.scss";

interface Props extends LinkProps {
  backgroundColor?: "default" | "white";
  icon: React.ReactElement;
  text?: string;
}

const IconLink: React.FC<Props> = ({
  backgroundColor = "default",
  icon,
  text,
  ...rest
}) => {
  return (
    <Link
      {...rest}
      className={classNames(styles.iconLink, styles[backgroundColor])}
    >
      <div className={styles.iconWrapper}>{icon}</div>
      {text && <div className={styles.textWrapper}>{text}</div>}
    </Link>
  );
};

export default IconLink;
