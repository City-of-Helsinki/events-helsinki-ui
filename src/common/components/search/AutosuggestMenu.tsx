import classNames from 'classnames';
import { IconCross } from 'hds-react';
import React, { FunctionComponent } from 'react';
import { useTranslation } from 'react-i18next';

import { AutosuggestMenuOption } from '../../types';
import styles from './autosuggestMenu.module.scss';

interface Props {
  focusedOption: number;
  isOpen: boolean;
  onClose: () => void;
  onOptionClick: (item: AutosuggestMenuOption) => void;
  options: AutosuggestMenuOption[];
}

const AutosuggestMenu: FunctionComponent<Props> = ({
  focusedOption,
  isOpen,
  onClose,
  onOptionClick,
  options,
}) => {
  const { t } = useTranslation();

  if (!isOpen) return null;

  return (
    <div className={styles.autosuggestMenu}>
      <div className={styles.title}>
        {t('commons.autosuggest.menu.title')}
        <button
          aria-label={t('commons.autosuggest.menu.ariaButtonClose')}
          className={styles.closeButton}
          onClick={onClose}
        >
          <IconCross />
        </button>
      </div>
      <ul className={styles.autosuggestOptions} role="listbox">
        {options.map((option, index) => {
          const handleClick = () => {
            onOptionClick(option);
          };

          return (
            <li
              key={index}
              className={classNames(
                styles.autosuggestOption,
                styles[`autosuggestOption--${option.type}`],
                {
                  [styles['autosuggestOption--isFocused']]:
                    focusedOption === index,
                }
              )}
              role="option"
              tabIndex={-1}
              onClick={handleClick}
              aria-selected={focusedOption === index}
            >
              <div className={styles.icon}></div>
              <div className={styles.textWrapper}>
                <div className={styles.text}>{option.text}</div>
              </div>
            </li>
          );
        })}
      </ul>

      <div className={styles.info}>{t('commons.autosuggest.menu.info')}</div>
    </div>
  );
};

export default AutosuggestMenu;
