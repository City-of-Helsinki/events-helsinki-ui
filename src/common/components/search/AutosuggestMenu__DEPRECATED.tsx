import classNames from 'classnames';
import { IconCross } from 'hds-react';
import React, { FunctionComponent } from 'react';
import { useTranslation } from 'react-i18next';

import { AutosuggestMenuOption__DEPRECATED } from '../../types';
import styles from './autosuggestMenu__DEPRECATED.module.scss';

interface Props {
  focusedOption: number;
  isOpen: boolean;
  onClose: () => void;
  onOptionClick: (item: AutosuggestMenuOption__DEPRECATED) => void;
  options: AutosuggestMenuOption__DEPRECATED[];
}

const AutosuggestMenu: FunctionComponent<Props> = ({
  focusedOption,
  isOpen,
  onClose,
  onOptionClick,
  options,
}) => {
  const { t } = useTranslation();

  if (!isOpen) return null;

  return (
    <div className={styles.autosuggestMenu}>
      <div className={styles.title}>
        {t('commons.autosuggest__DEPRECATED.menu.title')}
        <button
          aria-label={t('commons.autosuggest__DEPRECATED.menu.ariaButtonClose')}
          className={styles.closeButton}
          onClick={onClose}
        >
          <IconCross aria-hidden />
        </button>
      </div>
      <ul className={styles.autosuggestOptions} role="listbox">
        {options.map((option, index) => {
          const handleClick = () => {
            onOptionClick(option);
          };

          return (
            <li
              key={index}
              className={classNames(
                styles.autosuggestOption,
                styles[`autosuggestOption--${option.type}`],
                {
                  [styles['autosuggestOption--isFocused']]:
                    focusedOption === index,
                }
              )}
              role="option"
              tabIndex={-1}
              onClick={handleClick}
              aria-selected={focusedOption === index}
            >
              <div className={styles.text}>{option.text}</div>
            </li>
          );
        })}
      </ul>

      <div className={styles.info}>
        {t('commons.autosuggest__DEPRECATED.menu.info')}
      </div>
    </div>
  );
};

export default AutosuggestMenu;
