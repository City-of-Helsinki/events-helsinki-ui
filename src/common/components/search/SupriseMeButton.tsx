import React, { FunctionComponent } from "react";

import { ReactComponent as SupriseMeIcon } from "../../../assets/icons/svg/suprise-me.svg";
import { formatMessage } from "../../translation/TranslationUtils";
import styles from "./supriseMeButton.module.scss";

interface Props {
  onClick: () => void;
}

const SupriseMeButton: FunctionComponent<Props> = ({ onClick }) => {
  return (
    <button className={styles.supriseMeButton} onClick={onClick} type="button">
      <div className={styles.text}>{formatMessage("commons.supriseMe")}</div>
      <SupriseMeIcon />
    </button>
  );
};

export default SupriseMeButton;
