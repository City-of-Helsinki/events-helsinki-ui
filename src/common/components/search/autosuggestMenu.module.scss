@import "colors";

.autosuggestMenu {
  position: absolute;
  background: $hel-white;
  box-shadow: 0px 0px 8px 0px rgba(0, 0, 0, 0.2);
  min-height: 20px;
  right: 0;
  left: 0;
  top: 100%;
  z-index: 1;

  .title {
    padding: 1rem;
    text-align: center;
    color: $medium-gray;
    font-weight: 500;
    border-bottom: 1px solid $lightest-gray;

    .closeButton {
      border: none;
      background: transparent;
      padding: 0;
      position: absolute;
      right: 1rem;

      svg {
        display: block;
        margin: 0;
      }
      &:focus {
        outline: 2px solid $hel-black;
        outline-offset: 2px;
      }
    }
  }

  .info {
    padding: 1rem;
    text-align: center;
    color: $medium-gray;
    font-weight: 500;
  }

  .autosuggestOptions {
    padding: 0;
    margin: 0;
  }

  .autosuggestOption {
    position: relative;
    border-bottom: 1px solid $lightest-gray;
    display: flex;
    align-items: center;

    .colorIndicator {
      top: 0;
      bottom: -1px;
      position: absolute;
      width: 0.25rem;
      background-color: $hel-black;
    }

    .icon {
      height: 1.5rem;
      width: 1.5rem;
      margin-left: 1rem;

      svg {
        height: 1.5rem;
        width: 1.5rem;
      }
    }
    .textWrapper {
      flex: 1 1 0%;
      cursor: pointer;

      .text {
        padding: 0.875rem;
        color: $hel-black;
      }
    }
  }

  .autosuggestOption--isFocused {
    background-color: $hel-fog;
  }

  .autosuggestOption--district {
    .colorIndicator {
      background-color: $hel-metro;
    } 

    .icon {
      display: flex;
      align-items: center;
      justify-content: center;

      svg {
        margin: 0
      }
    }
  }

  .autosuggestOption--keyword {
    .colorIndicator {
      background-color: $hel-metro-light-50;
    }
  }

  .autosuggestOption--place {
    .colorIndicator {
      background-color: $hel-tram;
    } 
  }

  .autosuggestOption--yso {
    .colorIndicator {
      background-color: $hel-metro-light-20;
    } 
  }
}