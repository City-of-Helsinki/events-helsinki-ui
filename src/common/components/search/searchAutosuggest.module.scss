@import "colors";

.searchAutosuggest {
  position: relative;
  min-height: 3.375rem;
  background: $hel-white;
  display: flex;
  padding: 0 1rem;
  align-items: center;
  box-sizing: border-box;

  .iconWrapper {
    display: flex;
    align-self: center; 
  }

  svg {
    height: 1.25rem;
    width: 1.25rem;
    margin-right: 1rem;
  }
  

  .inputWrapper {
    flex: 1 1 0%;
    min-width: 50%;

    input {
      height: 1.25rem;
      line-height: 1.25rem;
      width: 100%;
      padding-left: 0;
      border: none;
      outline: none;
    }
  }
}
