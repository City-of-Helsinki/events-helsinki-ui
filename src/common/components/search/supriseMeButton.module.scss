@import "colors";
@import "layout";

.supriseMeButton {
  height: 6.25rem;
  width: 6.25rem;
  border-radius: 50%;
  position: absolute;
  top: -4.75rem;
  right: 2.5rem;
  overflow: hidden;
  background-color: $hel-suomenlinna;
  border: none;
  cursor: pointer;

  .text {
    font-size: 0.875rem;
    position: absolute;
    top: 1.875rem;
    width: 6.25rem;
    left: 0;
    font-weight: 600;
    color: $hel-black;
  }

  svg {
    position: absolute;
    bottom: 0;
    height: 3rem;
    width: 3rem;
    transform: translateX(-50%);
  }

  @include respond-below(sm) {
    height: 4.75rem;
    width: 4.75rem;
    top: -3.75rem;

    .text {
      font-size: 0.6875rem;
      top: 0.75rem;
      width: 3.125rem;
      left: 0.8125rem;
    }

    svg {
      height: 2.25rem;
      width: 2.25rem;
    }
  }
}