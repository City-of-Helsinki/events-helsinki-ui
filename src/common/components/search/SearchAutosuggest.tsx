import { IconSearch } from 'hds-react';
import React, { ChangeEvent } from 'react';

import {
  AUTOSUGGEST_KEYWORD_BLACK_LIST,
  AUTOSUGGEST_TYPES,
} from '../../../constants';
import {
  LinkedEventsSource,
  useKeywordListQuery,
} from '../../../generated/graphql';
import useDebounce from '../../../hooks/useDebounce';
import useKeyboardNavigation from '../../../hooks/useDropdownKeyboardNavigation';
import useLocale from '../../../hooks/useLocale';
import getLocalisedString from '../../../util/getLocalisedString';
import { AutosuggestMenuOption } from '../../types';
import AutosuggestMenu from './AutosuggestMenu';
import styles from './searchAutosuggest.module.scss';

export interface SearchAutosuggestProps {
  name: string;
  onChangeSearchValue: (value: string) => void;
  onOptionClick: (item: AutosuggestMenuOption) => void;
  placeholder: string;
  searchValue: string;
  source?: LinkedEventsSource;
}

const SearchAutosuggest: React.FC<SearchAutosuggestProps> = ({
  name,
  onOptionClick,
  onChangeSearchValue,
  placeholder,
  searchValue,
  source,
}) => {
  const locale = useLocale();
  const container = React.useRef<HTMLDivElement | null>(null);
  const input = React.useRef<HTMLInputElement | null>(null);
  const [isMenuOpen, setIsMenuOpen] = React.useState(false);
  const internalInputValue = useDebounce(searchValue, 300);

  const { data: keywordsData, loading: loadingKeywords } = useKeywordListQuery({
    skip: !internalInputValue,
    variables: {
      hasUpcomingEvents: true,
      pageSize: 5,
      text: internalInputValue,
      source,
    },
  });

  const [autoSuggestItems, setAutoSuggestItems] = React.useState<
    AutosuggestMenuOption[]
  >([]);

  const ensureMenuIsClosed = React.useCallback(() => {
    if (isMenuOpen) {
      setIsMenuOpen(false);
    }
  }, [isMenuOpen]);

  const ensureMenuIsOpen = React.useCallback(() => {
    if (!isMenuOpen) {
      setIsMenuOpen(true);
    }
  }, [isMenuOpen]);

  const {
    focusedIndex,
    setup: setupKeyboardNav,
    teardown: teardownKeyboardNav,
  } = useKeyboardNavigation({
    container: container,
    listLength: autoSuggestItems.length,
    onKeyDown: (event: KeyboardEvent) => {
      switch (event.key) {
        case 'ArrowUp':
        case 'ArrowDown':
          ensureMenuIsOpen();
          break;
        case 'Escape':
          handleCloseMenu();
          break;
        case 'Enter':
          const selectedItem = autoSuggestItems[focusedIndex];

          if (selectedItem) {
            handleMenuOptionClick(selectedItem);
          } else {
            // Search by text if no option is selected
            handleMenuOptionClick({
              text: searchValue,
              type: AUTOSUGGEST_TYPES.TEXT,
              value: searchValue,
            });
          }
          break;
        case 'Tab':
          ensureMenuIsClosed();
      }
    },
  });

  React.useEffect(() => {
    if (loadingKeywords) return;

    const items: AutosuggestMenuOption[] = [];
    const textItem = {
      text: internalInputValue,
      type: AUTOSUGGEST_TYPES.TEXT,
      value: internalInputValue,
    };

    items.push(textItem);

    items.push(
      ...(keywordsData?.keywordList.data
        .filter((keyword) => {
          const name = getLocalisedString(keyword.name, locale).toLowerCase();
          return (
            !AUTOSUGGEST_KEYWORD_BLACK_LIST.includes(keyword.id || '') &&
            name &&
            name !== textItem.text.toLowerCase()
          );
        })
        .map((keyword) => ({
          text: getLocalisedString(keyword.name, locale),
          type: AUTOSUGGEST_TYPES.KEYWORD,
          value: keyword.id || '',
        })) || [])
    );

    setAutoSuggestItems(items.filter((item) => item.text));
  }, [internalInputValue, keywordsData, loadingKeywords, locale]);

  const handleCloseMenu = React.useCallback(() => {
    setFocusToInput();

    ensureMenuIsClosed();
  }, [ensureMenuIsClosed]);

  const handleMenuOptionClick = React.useCallback(
    (option: AutosuggestMenuOption) => {
      onOptionClick(option);
      handleCloseMenu();
    },
    [handleCloseMenu, onOptionClick]
  );

  const setFocusToInput = () => {
    input.current?.focus();
  };

  const handleComponentClick = () => {
    setFocusToInput();
  };

  const onDocumentClick = (event: MouseEvent) => {
    const target = event.target;
    const current = container.current;

    // Close menu when clicking outside of the component
    if (!(target instanceof Node && current?.contains(target))) {
      setIsMenuOpen(false);
    }
  };

  const onDocumentFocusin = React.useCallback(
    (event: FocusEvent) => {
      const target = event.target;
      const current = container.current;

      if (!(target instanceof Node && current?.contains(target))) {
        ensureMenuIsClosed();
      }
    },
    [ensureMenuIsClosed]
  );

  const handleInputChange = (event: ChangeEvent<HTMLInputElement>) => {
    const newValue = event.target.value;

    onChangeSearchValue(newValue);

    ensureMenuIsOpen();
  };

  const handleInputFocus = () => {
    ensureMenuIsOpen();
  };

  React.useEffect(() => {
    setupKeyboardNav();
    document.addEventListener('click', onDocumentClick);
    document.addEventListener('focusin', onDocumentFocusin);
    // Clean up event listener to prevent memory leaks
    return () => {
      teardownKeyboardNav();
      document.removeEventListener('click', onDocumentClick);
      document.removeEventListener('focusin', onDocumentFocusin);
    };
  }, [onDocumentFocusin, setupKeyboardNav, teardownKeyboardNav]);

  return (
    <div
      onClick={handleComponentClick}
      className={styles.searchAutosuggest}
      ref={container}
    >
      <div className={styles.iconWrapper}>
        <IconSearch size="s" aria-hidden />
      </div>
      <div className={styles.inputWrapper}>
        <input
          ref={input}
          id={name}
          name={name}
          onChange={handleInputChange}
          onFocus={handleInputFocus}
          placeholder={placeholder}
          type="text"
          value={searchValue}
        />
      </div>
      <AutosuggestMenu
        focusedOption={focusedIndex}
        isOpen={!!searchValue && isMenuOpen}
        onClose={handleCloseMenu}
        onOptionClick={handleMenuOptionClick}
        options={autoSuggestItems}
      />
    </div>
  );
};

export default SearchAutosuggest;
