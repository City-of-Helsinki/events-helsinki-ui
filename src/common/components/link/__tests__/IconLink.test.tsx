import { shallow } from "enzyme";
import React from "react";
import { MemoryRouter } from "react-router";

import { ReactComponent as SearchIcon } from "../../../../assets/icons/svg/search.svg";
import IconLink from "../IconLink";

it("Navbar matches snapshot", () => {
  const container = shallow(
    <MemoryRouter>
      <IconLink icon={<SearchIcon />} text="test" to="/test" />
    </MemoryRouter>
  );
  expect(container.html()).toMatchSnapshot();
});
