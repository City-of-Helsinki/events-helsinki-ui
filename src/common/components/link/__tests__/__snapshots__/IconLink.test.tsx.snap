// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar matches snapshot 1`] = `"<div class=\\"iconLink\\"><a href=\\"/test\\"><div class=\\"iconWrapper\\"><svg>search.svg</svg></div><div class=\\"textWrapper\\">test</div></a></div>"`;
